edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes
+{
+    using System;
+    using System.Linq;
+    using System.Text;
+    using CoreConstants = Microsoft.Azure.Devices.Edge.Agent.Core.Constants;
+
+    public static class KubeUtils
+    {
+        const string Alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+        const string AllowedCharsLabelValues = "-._";
+        const string AllowedCharsDns = "-";
+        const string AllowedCharsGeneric = "-.";
+        const int MaxK8SValueLength = 253;
+        const int MaxDnsNameLength = 63;
+        const int MaxLabelValueLength = 63;
+
+        static bool IsAlpha(char ch) => Alphabet.IndexOf(ch) != -1;
+
+        static bool IsAlphaNumeric(char ch) => IsAlpha(ch) || Char.IsDigit(ch);
+
+        public static readonly string K8sNamespace = Constants.k8sNamespaceBaseName;
+
+        static KubeUtils()
+        {
+            // build the k8s namespace to which all things belong
+            string nsBaseName = Environment.GetEnvironmentVariable(Constants.k8sNamespaceBaseName);
+            string deviceId = Environment.GetEnvironmentVariable(CoreConstants.DeviceIdVariableName);
+            string hubName = Environment.GetEnvironmentVariable(CoreConstants.IotHubHostnameVariableName);
+            if (hubName != null && deviceId != null && nsBaseName != null)
+            {
+                hubName = hubName.Split('.').FirstOrDefault() ?? hubName;
+
+                K8sNamespace = $"{nsBaseName}-{hubName.ToLower()}-{deviceId.ToLower()}";
+            }
+        }
+
+        // Valid annotation keys have two segments: an optional prefix and name, separated by a slash (/). 
+        // The name segment is required and must be 63 characters or less, beginning and ending with an 
+        // alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. 
+        // The prefix is optional. If specified, the prefix must be a DNS subdomain
+        public static string SanitizeAnnotationKey(string key)
+        {
+            char[] annotationSplit = { '/' };
+            string[] keySegments = key.Split(annotationSplit, 2);
+            if (keySegments.Count() == 2)
+            {
+                return SanitizeDNSDomain(keySegments[0]) + "/" + SanitizeNameValue(keySegments[1]);
+            }
+            else
+            {
+                return SanitizeNameValue(key);
+            }
+        }
+
+        // Alphanumeric, '-' and '.' up to 253 characters.
+        public static string SanitizeK8sValue(string name)
+        {
+            name = name.ToLower();
+
+            var output = new StringBuilder();
+            for (int i = 0, len = 0; i < name.Length && len < MaxK8SValueLength; i++)
```

Comment:
Do we need to track length separately in `len`? Wouldn't `output.Length` give us the length?